### PR TITLE
Anchorsupdate

### PR DESCRIPTION
**Pull request by veanes**
First major integration of SRM into .NET Regex.
Includes support for anchors and initial unit tests.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtimelab?pullRequest=588) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtimelab?pullRequest=588) before we can accept your contribution.<br/><br/>:x: veanes [sign now](https://cla.dotnetfoundation.org/dotnet/runtimelab?pullRequest=588)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtimelab?pullRequest=588) it.</sub>

**Comment by veanes** [comment]
As agreed, I added a public RegexOtions.DFA enum that triggers SRM from within the regex Run method.

**Comment by veanes** [comment]
Did a major rewrite of the core SRM matcher engine, now with full anchor support. 
The code became much simpler and straightforward with a different (new) mechanism 
for handling of anchors. Still some unit tests fail and some bugs are to be fixed.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/srm/RegexOptions.cs [hunk 1]
```diff
@@ -1,7 +1,7 @@
 using System;
 using System.Diagnostics;
 
-namespace Microsoft.SRM
+namespace System.Text.RegularExpressions.SRM
 {
     [Serializable]
     internal struct RegexOptions
```
Comment (danmosemsft) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/srm/RegexOptions.cs:7]: Now that you're merging the code, seems the SRM versions of IgnoreCase, etc can be replaced by simply using the existing enum? And maybe retaining some of your helper methods here but no more need for any "mapping" from one to the other.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs [hunk 2]
```diff
@@ -113,9 +117,30 @@ private void Init(string pattern, RegexOptions options, TimeSpan matchTimeout, C
...
+            if (_useSRM)
+                // remove the DFA flag because it is undefined in SRM
+                _srm = InitializeSRM(tree.Root, options & ~RegexOptions.DFA);
+
             InitializeReferences();
         }
 
+        /// <summary>
+        /// This method is here for perf reasons: if InitializeSRM is NOT called in the
+        /// Init method, we don't load SRM.Regex when instantiating a regex that does not use the DFA option
```
Comment (danmosemsft) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs:133]: Now it's all in the same assembly, does no-inlining matter? The type will be loaded anyway, and Jitting will only happen if it ends up needing the code.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/srm/RegexOptions.cs (hunk 1 above)
Comment (veanes) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/srm/RegexOptions.cs]: Right, removed SRM.RegexOptions.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs (hunk 2 above)
Comment (veanes) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs]: Removed the comment and the complier method attribute. Added another note there as a reminder that in fact RightToLeft can be supported now -- fairly easily I think.

**Review by danmoseley — APPROVED** [review]
I didn't review the SRM parts, but looks good to me

**Comment by veanes** [comment]
This is great! Thanks!

**Comment by veanes** [comment]
I fixed the merge conflict in the project files.

**Comment by danmosemsft** [comment]
@veanes thanks, as you saw I did a big merge to bring everything up to date. That probably introduced the file header check as well.

**Pull request merged**